### PR TITLE
Multi-whitelisted operators [audited]

### PR DESCRIPTION
**Pull request by mtabasco**
The new feature allows operators to whitelist multiple EOAs/generic contracts and one external whitelisting contract.

**EOAs/generic contracts**
- Managed by the internal SSV whitelisting module
- An address can be whitelisted by many operators efficiently using bitmaps
- Backward compatible with the current contracts data

**Whitelisting contracts**
- One per operator
- Must implement the interface `ISSVWhitelistingContract` and comply with [ERC165](https://eips.ethereum.org/EIPS/eip-165)

### CODE REVIEW

**Review by andrew-blox — CHANGES_REQUESTED** [review]

**Review by andrew-blox — CHANGES_REQUESTED** [review]
There is no code nor tests dealing with whitelisted operators upon reactivation, this must be included

File: contracts/modules/SSVOperatorsWhitelist.sol [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        OperatorLib.checkZeroAddress(whitelistAddress);
+
+        if (OperatorLib.isWhitelistingContract(whitelistAddress))
+            revert AddressIsWhitelistingContract(whitelistAddress);
+
+        StorageData storage s = SSVStorage.load();
+        s.operators[operatorId].checkOwner();
+
+        if (!s.operators[operatorId].whitelisted) s.operators[operatorId].whitelisted = true;
+
```
[the remark this answers is not in the document]
Comment (andrew-blox) [reply contracts/modules/SSVOperatorsWhitelist.sol:29]: @mtabasco is was decided by @arielssv we will remove this automatic whitelisting in all places

File: contracts/modules/SSVOperatorsWhitelist.sol [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+            address currentWhitelisted = s.operatorsWhitelist[operatorId];
+
+            // operator already whitelisted? EOA or generic contract
+            if (currentWhitelisted != address(0)) {
+                (uint256 blockIndex, uint256 bitPosition) = OperatorLib.getBitmapIndexes(operatorId);
+                delete s.operatorsWhitelist[operatorId];
+                s.addressWhitelistedForOperators[currentWhitelisted][blockIndex] |= (1 << bitPosition);
+            } else {
+                operator.whitelisted = true;
+            }
```
[the remark this answers is not in the document]
Comment (andrew-blox) [reply contracts/modules/SSVOperatorsWhitelist.sol:82]: @mtabasco is was decided by @arielssv we will remove this automatic whitelisting in all places

File: contracts/modules/SSVOperatorsWhitelist.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (mtabasco) [reply contracts/modules/SSVOperatorsWhitelist.sol]: @andrew-blox Done. This change requires the operator owner to call the `setOperatorsPrivateUnchecked` function to make the operator private. If not, anyone can register validators using it even having whitelisted addresses.

File: contracts/modules/SSVOperatorsWhitelist.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (mtabasco) [reply contracts/modules/SSVOperatorsWhitelist.sol]: @andrew-blox Done. This change requires the operator owner to call the setOperatorsPrivateUnchecked function to make the operator private.
Also, the logic of whitelisted changed a bit. Thanks

File: contracts/libraries/CoreLib.sol [hunk 3]
```diff
@@ -1,13 +1,13 @@
...
+pragma solidity 0.8.24;
 
 import "./SSVStorage.sol";
 
 library CoreLib {
     event ModuleUpgraded(SSVModules indexed moduleId, address moduleAddress);
 
     function getVersion() internal pure returns (string memory) {
-        return "v1.1.0";
+        return "v1.2.0";
```
[the remark this answers is not in the document]
Comment (mtabasco) [reply contracts/libraries/CoreLib.sol:10]: No more places. Both SSVNetwork and SSVNetworkViews read this function.

File: contracts/modules/SSVOperators.sol [hunk 4]
```diff
@@ -62,6 +63,7 @@ contract SSVOperators is ISSVOperators {
         operator.snapshot.balance = 0;
         operator.validatorCount = 0;
         operator.fee = 0;
+        operator.whitelisted = false;
```
[the remark this answers is not in the document]
Comment (mtabasco) [reply contracts/modules/SSVOperators.sol]: I removed this line to keep the same last known privacy state when the operator is removed, so it can be fetched with `getOperatorById`

File: hardhat.config.ts [hunk 5]
```diff
@@ -1,14 +1,16 @@
-import 'dotenv/config';
-
 import { HardhatUserConfig } from 'hardhat/config';
 import { NetworkUserConfig } from 'hardhat/types';
-import '@nomicfoundation/hardhat-toolbox';
+
+import 'dotenv/config';
```
[the remark this answers is not in the document]
Comment (mtabasco) [reply hardhat.config.ts:4]: this version has upgrades on hardhat, clean-up dependencies, etc. If I remove it, it doesn't take `env` params properly. Maybe in the past it was imported in another place...

File: tasks/deploy.ts [hunk 6]
```diff
@@ -191,10 +185,10 @@ subtask('deploy:ssv-network', 'Deploys SSVNetwork contract')
         kind: 'uups',
       },
     );
-    await ssvNetwork.deployed();
+    await ssvNetwork.waitForDeployment();
 
-    const ssvNetworkProxyAddress = ssvNetwork.address;
-    const ssvNetworkImplAddress = await upgrades.erc1967.getImplementationAddress(ssvNetwork.address);
+    const ssvNetworkProxyAddress = await ssvNetwork.getAddress();
```
[the remark this answers is not in the document]
Comment (mtabasco) [reply tasks/deploy.ts:190]: yes, previously we used ethersjs for tests and tasks and now we use viem, which has a different syntax.

File: contracts/modules/SSVViews.sol [hunk 7]
```diff
@@ -49,13 +49,79 @@ contract SSVViews is ISSVViews {
...
+        validatorCount = operator.validatorCount;
+        whitelistedAddress = SSVStorage.load().operatorsWhitelist[operatorId];
+        isPrivate = operator.whitelisted;
+        isActive = operator.snapshot.block != 0;
+    }
+
+    function getWhitelistedOperators(
+        uint64[] calldata operatorIds,
+        address whitelistedAddress
+    ) external view override returns (uint64[] memory whitelistedOperatorIds) {
```
[the remark this answers is not in the document]
Comment (mtabasco) [reply contracts/modules/SSVViews.sol:81]: Done: `isAddressWhitelistedInWhitelistingContract`

**Comment by andrew-blox** [comment]
@mtabasco did you check the failing tests here?

File: tasks/deploy.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (andrew-blox) [reply tasks/deploy.ts]: ah cool also UI team switching from ethers to viem :)

File: contracts/modules/SSVOperatorsWhitelist.sol [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+            s.operatorsWhitelist[operatorId] = address(whitelistingContract);
+        }
+
+        emit OperatorWhitelistingContractUpdated(operatorIds, address(whitelistingContract));
```
[the remark this answers is not in the document]
Comment (andrew-blox) [reply contracts/modules/SSVOperatorsWhitelist.sol:83]: agreed

File: contracts/modules/SSVOperatorsWhitelist.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (andrew-blox) [reply contracts/modules/SSVOperatorsWhitelist.sol]: @arielssv i know you already know just affirming it here

File: contracts/modules/SSVViews.sol [hunk 9]
```diff
@@ -49,13 +50,88 @@ contract SSVViews is ISSVViews {
...
+        assembly {
+            mstore(whitelistedOperatorIds, count)
+        }
+    }
+
+    function isWhitelistingContract(address contractAddress) external view override returns (bool) {
+        return OperatorLib.isWhitelistingContract(contractAddress);
+    }
+
+    function isAddressWhitelistedInWhitelistingContract(
```
Comment (andrew-blox) [inline contracts/modules/SSVViews.sol:128]: The get that uses our contract can use array of operator ids and addresses, is it worth to make this the same or keep it single?

File: contracts/libraries/OperatorLib.sol [hunk 10]
```diff
@@ -30,67 +33,208 @@ library OperatorLib {
...
             }
+            cumulativeIndex += operator.snapshot.index;
 
             unchecked {
                 ++i;
             }
         }
     }
+
+    function updateMultipleWhitelists(
```
Comment (andrew-blox) [inline contracts/libraries/OperatorLib.sol:141]: We dont do any checking of duplicate addresses or duplicate operator ids, i dont see an issue with not checking this just making sure theres nothing maybe with bitmap or something else im not missing

File: contracts/modules/SSVOperatorsWhitelist.sol [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+            operator.checkOwner();
+
+            s.operatorsWhitelist[operatorId] = address(0);
+        }
+
+        emit OperatorWhitelistingContractUpdated(operatorIds, address(0));
+    }
+
+    function setOperatorsPrivateUnchecked(uint64[] calldata operatorIds) external override {
```
Comment (andrew-blox) [inline contracts/modules/SSVOperatorsWhitelist.sol:104]: chatted with @arielssv on this name, i dont like it lol, he likes it, just leaving a comment that i think its very confusing, i imagine that private has a checkmark and we are unchecking it ie making it public

Comment (mtabasco) [reply contracts/modules/SSVOperatorsWhitelist.sol:104]: @andrew-blox of course, all comments are welcomed and I'm all in to get a better name. You know the 
 meaning of the `unchecked` suffix: don't check if the operator has any whitelisted address.
`setOperatorsPrivateUnsafe` for me sounds too aggressive.
maybe just `setOperatorsPrivate`? The doc of the function says:
```
/// @notice Set the list of operators as private without checking for any whitelisting address
/// @notice The operators are considered private when registering validators
```
    cc @arielssv

File: contracts/libraries/OperatorLib.sol (hunk 10 above)
Comment (mtabasco) [reply contracts/libraries/OperatorLib.sol:141]: @andrew-blox for `setOperatorMultipleWhitelists` and `removeOperatorMultipleWhitelists` we require the list of operators to be unique and ordered. The check is performed in `OperatorLib.generateBlockMasks()` function.
In `OperatorLib.updateMultipleWhitelists` we check for empty `operatorIds` and `whitelistAddresses`.

In `setOperatorsWhitelistingContract` we check for a valid whitelisting contract and implicitly for empty address, also empty `operatorIds`. Same check for empty `operatorIds` in `removeOperatorsWhitelistingContract`, `setOperatorsPrivateUnchecked`, `setOperatorsPublicUnchecked `

It's true that in `setOperatorsWhitelistingContract`, `removeOperatorsWhitelistingContract`, `setOperatorsPrivateUnchecked`, and `setOperatorsPublicUnchecked` there is no check for duplicate `operatorIds`. This only implies extra gas costs for the users that pass unique lists and no risk for the ones who pass duplicates.

worth adding some tests on duplicates? I can do at the end of priorities

File: contracts/modules/SSVViews.sol (hunk 9 above)
Comment (mtabasco) [reply contracts/modules/SSVViews.sol:128]: @andrew-blox Do you mean `SSVViews.getWhitelistedOperators()`? that's for SSV whitelisting module addresses. To check many operatorIds and addresses in a external whitelisting contract, instead of implementing ourselves (just a loop of `ISSVWhitelistingContract(whitelistingContract).isWhitelisted(addressToCheck, operatorId)` calls, it's better to add a function to the `ISSVWhitelistingContract` interface and let the users implement it. Like:

`isWhitelisted(address[] addresses, uint64[] operatorIds)`

That's in the case we want to allow users to call external whitelisting contracts through us. External contracts can implement their functions to check the whitelisting status of accounts.

WDYT?

File: contracts/libraries/OperatorLib.sol (hunk 10 above)
Comment (andrew-blox) [reply contracts/libraries/OperatorLib.sol:133]: I think its worth to add a duplicate test but not important

**Review by andrew-blox — APPROVED** [review]

**Review by Mohsen-T — APPROVED** [review]
I carefully reviewed every modified or added code, and everything looks good to merge into the branch

**Review by lior-blox — APPROVED** [review]
